### PR TITLE
Use cudf like operator in GpuLike operator

### PR DESCRIPTION
**Pull request by cindyyuanjiang**
The **problem** was raised in issue #6430, suggesting to switch over to the cudf `like` operator for `GpuLike` operator.

We performed the change and benchmarked the performance of the new `GpuLike` operator (using cudf `like`) against the existing one (using cudf `matches_re`). 

We run the new and existing `GpuLike` operator with equivalent like patterns over the same datasets. The benchmark test was set up so that both functions (`like` vs `matches_re`) are required to read the entire matching string to resolve to true/false. This provides fair comparison between the different implementations. The results are shown below varying 0.1M to 1B rows with 1-100% hit-rates (percentage of matches in the datasets). The y-axis shows the speedup of the **new** cudf `like` operator against cudf `matches_re` operator as an x-factor, ranged between 2x to 4x. 

perf

We observed that across all benchmarks, `matches_re` uses 1.3x to 1.6x total GPU memory comparing against `like`.

### CODE REVIEW

**Comment by cindyyuanjiang** [comment]
We computed the speedup of `like` against `matches_re` varying string lengths from 4 to 128 in a column.

strlen1
strlen2

In addition, we computed the speedup of `like` against `matches_re` varying the percentages of matchings/comparisons (1-100%) that require the GpuLike operator to read the entire string to resolve to true (e.g. patterns like "a_%_%" does not require reading the entire matching string).
match_percentage

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala [hunk 1]
```diff
@@ -719,8 +719,7 @@ case class GpuLike(left: Expression, right: Expression, escapeChar: Char)
     } else {
       null
     }
-    val regexStr = escapeLikeRegex(likeStr, escapeChar)
-    lhs.getBase.matchesRe(regexStr)
+    lhs.getBase.like(Scalar.fromString(likeStr), Scalar.fromString(Character.toString(escapeChar)))
```
Comment (revans2) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala:722]: The scalars are leaked here. We need to be sure that we can close them, also GpuScalar already holds a Scalar so we can use it directly.

```
override def doColumnar(lhs: GpuColumnVector, rhs: GpuScalar): ColumnVector = {
  withResource(Scalar.fromString(Character.toString(escapeChar))) { escapeScalar =>
    lhs.getBase.like(rhs.getBase, escapeScalar)
  }
}
```

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala [hunk 2]
```diff
@@ -719,8 +719,7 @@ case class GpuLike(left: Expression, right: Expression, escapeChar: Char)
     } else {
       null
     }
-    val regexStr = escapeLikeRegex(likeStr, escapeChar)
```
Comment (revans2) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala:722]: We should remove escapeLikeRegex entirely. The new code appears to be better in every way.

**Comment by revans2** [comment]
@cindyyuanjiang 
These are really great numbers. Would it be possible to use your script to compare this to the CPU implementation as well? I don't think we have ever done that. Also could you post the script/query you used? I am really curious to know if this is the end to end query time or if you went in and looked at the time spent doing these particular operations using a trace or our metrics. There can be a lot of overhead in a query and if these speedups are for end to end query time, wow.

**Comment by cindyyuanjiang** [comment]
> 

Thank you very much @revans2! I will look into comparing to the CPU implementation now. For the current performance measurement, I compared the runtime of the `libcudf` operations using `nsys` under NVTX Range Summary. I will post my queries for reference. :)

**Comment by cindyyuanjiang** [comment]
I put my queries and some documentation on how I got the performance numbers in this [doc](https://docs.google.com/document/d/1jcKyxyAPJqXF9l1PbVl5pOlmU63IR3HwkkZ5uzA3SOI/edit?usp=sharing). Please leave a comment if you find anything unclear. Thanks!

**Comment by cindyyuanjiang** [comment]
I compared the performance of GPU like and matches_re implementation with the CPU implementations. This benchmark uses end-to-end query time of 200 like queries. The y-axis shows the speedup of the GPU matches_re and like against CPU implementation and the GPU like against GPU matches_re implementation

gpucpuquery
gpugpuquery

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala (hunk 1 above)
Comment (cindyyuanjiang) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala:722]: Updated. Thank you very much!

**Review by revans2 — APPROVED** [review]

**Comment by cindyyuanjiang** [comment]
build

**Pull request merged**